toolkit/components/filewatcher/NativeFileWatcherLnx.cpp

```diff
@@ -0,0 +1,1073 @@
+    mWatchedResourcesByPath;
+    nsDataHashtable<nsUint32HashKey, WatchedResourceDescriptor*>
+    mWatchedResourcesByHandle;
+
+    // The same callback can be associated to multiple watches so we need to keep
+    // them alive as long as there is a watch using them. We create two hashtables
+    // to map directory names to lists of nsMainThreadPtr<callbacks>.
+    nsClassHashtable<nsStringHashKey, ChangeCallbackArray> mChangeCallbacksTable;
+    nsClassHashtable<nsStringHashKey, ErrorCallbackArray> mErrorCallbacksTable;
+
+    // Other methods need to know that a shutdown is in progress.
+    bool mShuttingDown;
+
+    // Main inotify file descriptor (initialized in NativeFileWatcher's Init())
+    int mInotifyFileDescriptor;
+
+    // Queue for the events read from inotify file descriptor
+    std::queue<inotify_event*> mInotifyEventQueue;
+
+    nsresult RunInternal();
+
+    nsresult DispatchChangeCallbacks(
+            WatchedResourceDescriptor* aResourceDescriptor,
+            const nsAString& aChangedResource);
+
+    nsresult ReportChange(
+            const nsMainThreadPtrHandle<nsINativeFileWatcherCallback>& aOnChange,
+            const nsAString& aChangedResource);
+
+    nsresult DispatchErrorCallbacks(
+            WatchedResourceDescriptor* aResourceDescriptor,
+            nsresult anError,
+            long anOSError);
+
+    nsresult ReportError(
+            const nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>& aOnError,
+            nsresult anError,
+            long anOSError);
+
+    nsresult ReportSuccess(const nsMainThreadPtrHandle<
+                           nsINativeFileWatcherSuccessCallback>& aOnSuccess,
+                           const nsAString& aResourcePath);
+
+    void AppendCallbacksToHashtables(
+            const nsAString& aPath,
+            const nsMainThreadPtrHandle<nsINativeFileWatcherCallback>& aOnChange,
+            const nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>& aOnError);
+
+    void RemoveCallbacksFromHashtables(
+            const nsAString& aPath,
+            const nsMainThreadPtrHandle<nsINativeFileWatcherCallback>& aOnChange,
+            const nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>& aOnError);
+
+    nsresult MakeResourcePath(WatchedResourceDescriptor* changedDescriptor,
+                              const nsAString& resourceName,
+                              nsAString& nativeResourcePath);
+};
+
+/**
+ * The watching thread logic.
+ *
+ * @return NS_OK if the watcher loop must be rescheduled, a failure code
+ *         if it must not.
+ */
+nsresult
+NativeFileWatcherIOTask::RunInternal()
+{
+    int mInotifyReadTimeout;
+    char inotifyEventTempBuffer[(sizeof(struct inotify_event) + NAME_MAX + 1)] __attribute__ ((aligned(8)));
+
+    // Sleep to keep thread in suspended state until event occurs, and to keep thread from using CPU.
+    // Will return early as soon as an event occurs (our signal handler is triggered).
+    sleep(0.01);
```

Comment:
No magic constants, so please make this `0.01` a global `const SLEEP_DELAY_MS` or something such. Also is that [this `sleep`](https://linux.die.net/man/3/sleep)? The argument is an integer number of seconds, so what you're doing there is rounded to `sleep(0)`, which sounds odd.

Reply:
It looks like usleep is eventually deprecated by nanosleep from  <time.h>. 
More details here: https://linux.die.net/man/2/nanosleep

Reply:
This will also be removed in a blocking read implementation, preventing the need to sleep between signals and events.